client/v3/snapshot/v3_snapshot.go

```diff
@@ -54,38 +55,58 @@ func SaveWithVersion(ctx context.Context, lg *zap.Logger, cfg clientv3.Config, d
 	if err != nil {
 		return "", err
 	}
-	defer cli.Close()
+	defer func() {
+		err = cli.Close()
+		if err != nil {
+			lg.Error("Failed to close client", zap.Error(err))
+		}
+	}()
 
 	partpath := dbPath + ".part"
-	defer os.RemoveAll(partpath)
+	defer func() {
+		err = os.RemoveAll(partpath)
+		if err != nil {
+			lg.Error("Failed to cleanup .part file", zap.Error(err))
+		}
+	}()
 
-	var f *os.File
-	f, err = os.OpenFile(partpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fileutil.PrivateFileMode)
+	f, err := os.OpenFile(partpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fileutil.PrivateFileMode)
 	if err != nil {
 		return "", fmt.Errorf("could not open %s (%w)", partpath, err)
 	}
+	defer func() {
+		err = f.Close()
+		if err != nil && !errors.Is(err, os.ErrClosed) {
```

Comment:
It works. But I think a better and more explicit way is to use `sync.Once` to ensure f.Close is only executed once. So that it's well aware that we need to close the file before renaming, but also need to ensure it's closed in case early return due to any error.

Reply:
Checked https://pkg.go.dev/os#File.Close and the api can guarantee it can be executed only once. `sync.Once` is not neccesary here.

Reply:
Not a big deal, please feel free to merge this PR and backport to 3.5 and 3.4.